citadel/indico_citadel/search.py

```diff
@@ -0,0 +1,139 @@
+# This file is part of the Indico plugins.
+# Copyright (C) 2002 - 2021 CERN
+#
+# The Indico plugins are free software; you can redistribute
+# them and/or modify them under the terms of the MIT License;
+# see the LICENSE file for more details.
+
+import re
+
+import requests
+from werkzeug.urls import url_join
+
+from indico.modules.search.base import IndicoSearchProvider, SearchTarget
+
+
+class CitadelProvider(IndicoSearchProvider):
+    def __init__(self, *args, **kwargs):
+        from indico_citadel.plugin import CitadelPlugin
+
+        super().__init__(*args, **kwargs)
+        self.token = CitadelPlugin.settings.get('search_backend_token')
+        self.backend_url = CitadelPlugin.settings.get('search_backend_url')
+        self.records_url = url_join(self.backend_url, 'api/records/')
+
+    def search(self, query, access, object_type=SearchTarget.event, page=1, params=None, highlight=True):
+        # https://cern-search.docs.cern.ch/usage/operations/#query-documents
+        # this token is used by the backend to authenticate and also to filter
+        # the objects that we can actually read
+        headers = {
+            'Authorization': f'Bearer {self.token}'
+        }
+
+        filter_query, ranges = format_filters(params)
+        # Look for objects matching the `query` and schema, make sure the query is properly escaped
+        # https://cern-search.docs.cern.ch/usage/operations/#advanced-queries
+        q = '{} {} +type:{}'.format(format_query(query), ranges, object_type.name)
+        search_params = {'page': page, 'size': self.RESULTS_PER_PAGE, 'q': q, 'highlight': '_data.*', **filter_query}
+        # Filter by the objects that can be viewed by users/groups in the `access` argument
+        if access:
+            search_params['access'] = ','.join(access)
+
+        resp = requests.get(self.records_url, params=search_params, headers=headers, verify=False)
+        if not resp.ok:
+            return 0, [], []
+
+        resp = resp.json()
+        _aggregations, hits = resp['aggregations'], resp['hits']
+        total, _objects = hits['total'], hits['hits']
+
+        aggregations = [
+            {
+                'key': key,
+                'label': filters[key],
+                'buckets': value['buckets']
+            }
+            for key, value in _aggregations.items()
+            if len(key) and key in filters
```

Comment:
```suggestion
            if key and key in filters
```

`key and` is probably not needed

Reply:
If I recall correctly this was intended to prevent empty key aggregations, in cases where an affiliation might be empty and one of the filters might be `''` or (nothing).

Reply:
Right, this was intended to prevent empty aggregations. A few iterators later I added the filter whitelist...